### PR TITLE
soc: nordic: handle -include for IAR

### PR DESCRIPTION
**Pull request by RobinKastberg**
CMakeLists.txt uses the C compiler parameter -include, for IAR this is --preinclude.

### CODE REVIEW

**Review by masz-nordic — DISMISSED** [review]

File: soc/nordic/CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,11 @@ zephyr_library_sources(
 # headers for different SoCs.
 set(dt_binding_includes ${DTS_INCLUDE_FILES})
 list(FILTER dt_binding_includes INCLUDE REGEX "/dt-bindings/.*\.h$")
-list(TRANSFORM dt_binding_includes PREPEND "-include;")
+if(CMAKE_C_COMPILER_ID STREQUAL "IAR")
```
Comment (tejlmand) [inline soc/nordic/CMakeLists.txt:20]: we should avoid compiler / toolchain specific handling inside the code as it makes support for multiple toolchains much harder.

Instead the code itself should be cleaned up and the include flag should be defined in the compiler properties and picked up where used.

For example like this:

```
set(include_flag $<TARGET_PROPERTY:compiler,include-file>)

set_source_files_properties(
  validate_binding_headers.c
  DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
  PROPERTIES COMPILE_OPTIONS "${include_flag}$<JOIN:${dt_binding_includes},;${include_flag}>"
)
```

and then in https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/gcc/compiler_flags.cmake add a line like this:
```
set_compiler_property(PROPERTY include-file "-include$<SEMICOLON>")
# Note to @RobinKastberg: check if "-include;" is sufficient in above.
```

and in the correspong IAR `compiler_flags.cmake` you can then do similar:
```
set_compiler_property(PROPERTY include-file "-preinclude$<SEMICOLON>")
```
and there will no longer be any need for doing compiler check in the CMake code itself, plus a benefit is that another toolchain can just provide its own flag and doesn't have to add an extra if-clause to this CMake code.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this cleanup, but please make the changes to the toolchain / compiler infrastructure instead.

File: soc/nordic/CMakeLists.txt [hunk 2]
```diff
@@ -17,11 +17,14 @@ zephyr_library_sources(
 # headers for different SoCs.
 set(dt_binding_includes ${DTS_INCLUDE_FILES})
 list(FILTER dt_binding_includes INCLUDE REGEX "/dt-bindings/.*\.h$")
-list(TRANSFORM dt_binding_includes PREPEND "-include;")
+
+get_target_property(include_flag compiler include_file)
+set(include_flag ${include_flag}$<SEMICOLON>)
```
Comment (tejlmand) [inline soc/nordic/CMakeLists.txt:22]: please switch to using a genex. genex'es are evaluated at generator time whereas `get_target_property()` is evaluated at configure time. Thus genex is a bit safer, and should be used when possible.

```suggestion
set(include_flag $<TARGET_PROPERTY:compiler,include_file>$<SEMICOLON>)
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
looks good, a minor suggestion to keep a general safer pattern.
(devs tends to copy-paste code they find elsewhere)

**Review by tejlmand — APPROVED** [review]

**Comment by kartben** [comment]
@RobinKastberg this need rebasing to get past the unrelated CI failure

**Review by nashif — APPROVED** [review]

**Pull request merged**